File: Java/src/main/java/QuestionBank/Solution101_200/Solution107.java
package QuestionBank.Solution101_200;

import tool.TreeNode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author windows
 */
public class Solution107 {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        Map<Integer, List<Integer>> map = new HashMap<>();
        dfs(root, 0, map);
        ArrayList<List<Integer>> res = new ArrayList<>();
        for (int i = map.size() - 1; i >= 0; i--) {
            res.add(map.get(i));
        }
        return res;
    }

    private void dfs(TreeNode root, int level, Map<Integer, List<Integer>> map){
        if (root == null){
            return;
        }
        List<Integer> list = map.getOrDefault(level, new ArrayList<Integer>());
        list.add(root.val);
        map.put(level, list);
        dfs(root.left, level + 1, map);
        dfs(root.right, level + 1, map);
    }


    public static void main(String[] args) {
        Solution107 solution107 = new Solution107();
    }
}

File: C++/src/Solution101_200/Solution107.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution107.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
#include <map>
#include <functional>

using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


vector<vector<int>> levelOrderBottom(TreeNode* root) {
    vector<vector<int>> ans;
    map<int, vector<int>> hash;

    function<void(TreeNode*, int)> dfs = [&](TreeNode* node, int depth){
        if (node == nullptr){
            return;
        }
        hash[depth].push_back(node->val);
        dfs(node->left, depth + 1);
        dfs(node->right, depth + 1);
    };
    dfs(root, 0);
    for (const auto &item: hash){
        ans.push_back(item.second);
    }
    reverse(ans.begin(), ans.end());
    return ans;
}
